plugin/rollback.go

```diff
@@ -0,0 +1,121 @@
+package plugin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	rotateCredentialWAL = "rotateCredentialWAL"
+)
+
+// rotateCredentialEntry is used to store information in a WAL that can retry a
+// credential rotation in the event of partial failure.
+type rotateCredentialEntry struct {
+	LastVaultRotation  time.Time `json:"last_vault_rotation"`
+	LastPassword       string    `json:"last_password"`
+	CurrentPassword    string    `json:"current_password"`
+	RoleName           string    `json:"name"`
+	ServiceAccountName string    `json:"service_account_name"`
+	TTL                int       `json:"ttl"`
+}
+
+func (b *backend) walRollback(ctx context.Context, req *logical.Request, kind string, data interface{}) error {
+	switch kind {
+	case rotateCredentialWAL:
+		return b.handleRotateCredentialRollback(ctx, req.Storage, data)
+	default:
+		return fmt.Errorf("unknown WAL entry kind %q", kind)
+	}
+}
+
+func (b *backend) handleRotateCredentialRollback(ctx context.Context, storage logical.Storage, data interface{}) error {
+	var wal rotateCredentialEntry
+	if err := mapstructure.WeakDecode(data, &wal); err != nil {
+		return err
+	}
+
+	if wal.CurrentPassword == "" {
+		b.Logger().Warn("WAL does not contain a password for service account")
+		return nil
+	}
+
+	// Check creds for deltas. Exit if creds and WAL are the same.
+	path := fmt.Sprintf("%s/%s", storageKey, wal.RoleName)
+	credEntry, err := storage.Get(ctx, path)
+	if err == nil && credEntry != nil {
```

Comment:
If I understand this correctly, if there is an error retrieving the entry from storage this will not error here but instead error below when it tries to write to storage. Am I interpreting this correctly? If so, should this error here instead or is that redundant?